components/layout/query.rs

```diff
@@ -527,21 +530,58 @@ impl FragmentBorderBoxIterator for UnioningFragmentScrollAreaIterator {
 // https://drafts.csswg.org/cssom-view/#extensions-to-the-htmlelement-interface
 impl FragmentBorderBoxIterator for ParentOffsetBorderBoxIterator {
     fn process(&mut self, fragment: &Fragment, level: i32, border_box: &Rect<Au>) {
+        match self.node_offset_box {
+            Some(_) => {  }, // We've already found the node, so we already have its parent.
+            None => {
+                // Remove all nodes at this level or higher, as they can't be parents of this node.
+                self.parent_nodes.truncate(level as usize);
+                assert!(self.parent_nodes.len() == level as usize);
+            },
+        }
+
         if fragment.node == self.node_address {
             // Found the fragment in the flow tree that matches the
             // DOM node being looked for.
-            self.has_found_node = true;
-            self.node_border_box = *border_box;
+            self.has_processed_node = true;
+            self.node_offset_box = Some(NodeOffsetBoxInfo {
+                offset: border_box.origin,
+                rectangle: *border_box,
+            });
 
             // offsetParent returns null if the node is fixed.
             if fragment.style.get_box().position == computed_values::position::T::fixed {
                 self.parent_nodes.clear();
             }
-        } else if level > self.last_level {
+        } else if fragment.contains_node(self.node_address) {
+            // Found a fragment in the flow tree that matches a child
+            // of the DOM node being looked for.
+            // Since we haven't found an exact match, the node must be
+            // inline. Hopefully.
```

Comment:
@emilio Yes, this entire thing is truly horrible and relies on a lot of unenforced invariants. You could break it from some entirely separate part of layout, and the tests probably wouldn't tell you.

A much better way of handling this particular section has occurred to me, using `InlineFragmentNodeFlags`. In fact, the current strategy fails when applied to an element containing an {ib} split (never mind that KHTML/WebKit/Blink-based browsers all fail in such a situation as well!). I'm afraid this code hasn't aged quite as well as I thought, but no matter. I'll fix it.